File: reference/buergi_chase_matlab/compare_chase.py
"""
Cross-validation script: run our Python CHASE model (models/chase.py) on the
same fixed trial sequence and parameters as compare_chase.m, to check the
two implementations produce identical trial-by-trial outputs.

Run from anywhere: python3 reference/buergi_chase_matlab/compare_chase.py
"""

import os
import sys
import numpy as np

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", ".."))
from models.chase import CHASEModel

# 0-indexed: Rock=0, Paper=1, Scissors=2 (matches MATLAB's 1=Rock,2=Paper,3=Scissors minus 1)
choices     = np.array([0, 1, 2, 0, 1, 2, 0, 0, 1, 2, 1, 0])
opp_choices = np.array([1, 1, 1, 0, 0, 0, 2, 2, 2, 1, 0, 2])

params = {"alpha": 0.3, "beta": 3.0, "gamma": 2.0, "lam": 1.5, "kappa": 3}

model  = CHASEModel()
result = model.simulate(params, choices, opp_choices)

lik = result.action_probs[np.arange(len(choices)), choices]

np.set_printoptions(precision=4, suppress=True)

print("lik (P(chosen action) per trial):")
print(lik)
print()
print("SV (subjective value per trial):")
print(result.choice_values)
print()
print("APE (action prediction error per trial):")
print(result.ape)
print()
print("KL_div / BU (belief update per trial):")
print(result.belief_updates)
print()
print("beliefs (POSTERIOR after each trial, rows=trials, cols=levels 0..kappa-1):")
print(result.beliefs)

File: reference/buergi_chase_matlab/compare_chase.m
% Cross-validation script: run the original CHASE model on a fixed trial
% sequence with fixed parameters, to compare trial-by-trial outputs against
% the Python reimplementation in models/chase.py.
%
% Run this in MATLAB Online with the uploaded `source/` folder (and its
% MERLIN_toolbox subfolder) on the path.

addpath(genpath('source'));

% ── Model config: CH architecture, kappa fitted up to 3, RW-freq learning ──
model = BAKR_2024_CHASE_config('CH','fitted',3,'RW-freq');
disp('Parameter order:');
disp({model.params.name});

% ── Fixed trial sequence (1=Rock, 2=Paper, 3=Scissors) ─────────────────────
data.choice_own   = [1;2;3;1;2;3;1;1;2;3;2;1];
data.choice_other = [2;2;2;1;1;1;3;3;3;2;1;3];
data.missing      = zeros(12,1);
data.n_trials     = 12;
data.trial        = (1:12)';  % no block resets mid-sequence

% ── Fixed parameters, in the order printed above: beta, lambda, gamma, alpha, kappa ──
curr_params = [3.0; 1.5; 2.0; 0.3; 3];

[negLL, states, data] = BAKR_2024_CHASE_model(data, model, curr_params, 'fit');

fprintf('negLL: %.6f\n\n', negLL);

disp('lik (P(chosen action) per trial):');
disp(states.lik');

disp('SV (subjective value per trial):');
disp(states.subj_SV');

disp('APE (action prediction error per trial):');
disp(states.subj_APE');

disp('KL_div / BU (belief update per trial):');
disp(states.subj_KL_div');

disp('beliefs (PRIOR going into each trial, rows=trials, cols=levels 0..kappa-1):');
disp(states.beliefs);
